salvo/src/lib/benchmark/fully_dockerized_benchmark.py

```diff
@@ -24,25 +24,132 @@ def __init__(self, job_control: JobControl, benchmark_name: str) -> None:
     super(Benchmark, self).__init__(job_control, benchmark_name)
 
   def _validate(self) -> None:
-    """Validate that all data required for running a benchmark exist.
+    """Validate that all data required for running a benchmark exists.
+
+    Verify that all required images are present in the control object.
+    If not, verify that sources exist from which we can build the
+    required docker images.
 
     Returns:
         None
     """
-    pass
+    verify_source = False
+    images = self.get_images()
+
+    # Determine whether we need to build the images from source
+    # If so, verify that the required source data is defined
+    verify_source = images is None or \
+        not images.nighthawk_benchmark_image or \
+        not images.nighthawk_binary_image or \
+        not images.envoy_image
+
+    log.debug(f"Source verification needed: {verify_source}")
+    if verify_source:
+      self._verify_sources(images)
+
+    return
 
   def _verify_sources(self, images: DockerImages) -> None:
     """Validate that sources are available to build a missing image.
 
+    Verify that a source definition exists tht can build a missing
+    image needed for the benchmark.
+
     Returns:
         None
+
+    Raises:
+        an Exception if no source definitions allow us to build
+          missing docker images.
     """
-    pass
+    source = self.get_source()
+    if not source:
+      raise Exception("No source configuration specified")
```

Comment:
Typically it is a good practice to define our own exception for each module. Something like:

```python
class Error(Exception):
  """Raised on errors in this module."""
```

This way the callers of functions can choose to refer to our exception in `try->except` statements. Without this, they can only choose to catch `Exception` itself which covers pretty much any error in Python.

Applies to multiple locations.

Reply:
This comment doesn't seem to be addressed yet.

Reply:
I've created https://github.com/envoyproxy/envoy-perf/issues/82 to address this.

I do have a note to do this exact thing, however since this PR is getting large, I'd like to defer this to the next one.

Reply:
That is a very good point, let's leave it to a separate PR to keep them short. Thank you for suggesting it.